feat: add Litestar framework support and related middleware

pythonhubdev opened the pull request:
## Summary of change

(A few sentences about this PR)

## Related issues
-  #310

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [x] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.
-   [ ] If access token structure has changed
    -   Modified test in `tests/sessions/test_access_token_version.py` to account for any new claims that are optional or omitted by the core

## Remaining TODOs for this PR
- [x] Add `framework/litestar/__init__.py`
- [x] Add `litestar` to
  - [x] `setup.py`
  - [x] `Makefile`
- [ ] Support litestar in website tests
  - [x] Setup `tests/frontendIntegration/litestar`
  - [ ] Ensure `supertokens-website` tests run with litestar
- [ ] Support litestar in auth-react tests 
  - [x] Setup `tests/auth-react/litestar`
  - [ ] Ensure `supertokens-auth-react` tests run with litestar

pythonhubdev commented:
Hi @rishabhpoddar  @namsnath  I have made some changes to support litestar but the tests are failing need some help here since I couldn't debug where the issue is exactly coming from can any one guide me through what am I doing wrong?

namsnath commented:
Hey @pythonhubdev ,
Which test suite has the failures? Any specific tests?

I can take a look at this in detail sometime next week.

pythonhubdev commented:
Hi @namsnath thanks for the response so except the top  tests all other tests are failing. Mainly the tests that are calling the internal API of supertokens are failing but they are passing in FastAPI I don't know if I have to mount / attach the app or something so mostly it's 404 a little guidance help me I can fix it and the other issues are 500 Internal Server Error I don't know exactly why it's failing.

pythonhubdev commented:
Hi @namsnath I have updated the test and modified the Litestar middleware to use `ASGIMiddleware` as well but the routes related to supertoken i.e all the routes following `/auth` are **NOT FOUND**. Can you help me in fixing this even if you guide me where I have to look into I can take it from there. Thanks

pythonhubdev commented:
Hi @namsnath the major issue is I was not able to access those routes even manually not alone in testing I think the app is not getting attached or something is getting missed out due to some reason I can't access the `/auth/dasboard` route even. Can you guide me a little more please?

namsnath commented:
@pythonhubdev ,
I looked into this and found a few interesting issues.
- The requests to SuperTokens routes are not defined in the `route_handlers`, and result in a 404.
- Litestar handles 404/405s at the app level, and thus these do not flow down to the middleware and are handled directly through the `exception_handlers` map.
- `exception_handlers` can not be async functions, so it does not seem possible to call the ST middleware through these and handle the request.

I could get the requests routed correctly with these changes in the example app:
```python
# supertokens_python/framework/litestar/litestar_request.py
class LitestarRequest(BaseRequest):
    ....
    def get_path(self) -> str:
        return self.request.scope["raw_path"].decode("utf-8")

# main.py

# TODO: Move to `litestar_middleware.py`
# If using the import from `litestar_middleware` does not work, convert to a factory function
@asgi("/auth", is_mount=True, copy_scope=True)
async def supertokens_app(scope: Scope, receive: Receive, send: Send) -> None:
    # <code from `LitestarMiddleware.handle` without `next_app` handling>
    # Additional changes:
    asgi_response = result.response.to_asgi_response(app=None, request=request) # Removed `await`, Set `app=None`

app = Litestar(
    route_handlers=[supertokens_app, ...],
    # middleware=[],
    exception_handlers={...},
)
```

A few things to keep in mind for this PR:
- Code needs to work on Python 3.8
- Please set up hooks by running `make set-up-hooks`
- Try to make sure types are correctly mentioned, and avoid `Any` wherever possible

I'll add a few items to the TODO section of the PR as well.

### Ref:
- https://docs.litestar.dev/2/usage/exceptions.html#exception-handling-layers
- https://docs.litestar.dev/2/usage/routing/overview.html#mounting-asgi-apps
- https://github.com/litestar-org/litestar/issues/175
- https://github.com/litestar-org/litestar/issues/2262

namsnath commented on supertokens_python/utils.py:
```diff
@@ -45,6 +45,8 @@
 from supertokens_python.framework.response import BaseResponse
 from supertokens_python.logger import log_debug_message
 
+from .framework.litestar.framework import LitestarFramework
```
Use absolute imports

namsnath commented on supertokens_python/framework/litestar/litestar_request.py:
Rename to `litestar/request.py`

namsnath commented on supertokens_python/framework/litestar/litestar_response.py:
Rename to `litestar/response.py`

namsnath commented on tests/Litestar/__init__.py:
Rename the directory to be `litestar` instead of `Litestar`

namsnath commented on tests/Litestar/test_litestar.py:
```diff
@@ -0,0 +1,1127 @@
+    @get("/login")
+    async def login(request: Request[Any, Any, Any]) -> dict[str, Any]:
+        user_id = "userId"
+        await create_new_session(
+            request,
+            "public",
+            RecipeUserId(user_id),
+            {},
+            {},
+        )
```
Use keyword args for clarity:
```python
await create_new_session(request=request, ...)
```

namsnath commented on tests/Litestar/test_litestar.py:
```diff
@@ -0,0 +1,1127 @@
+            website_domain="http://supertokens.io",
+            api_base_path="/auth",
+        ),
+        framework="flask",
```
```suggestion
        framework="litestar",
```
Please fix all similar instances

namsnath commented on tests/Litestar/test_litestar.py:
```diff
@@ -0,0 +1,1127 @@
+            return Response(content={"message": "Unauthorized"}, status_code=401)
+
+    @get("/info")
+    async def info_get(request: Request[Any, Any, Any]) -> dict[str, Any]:
```
Use `Dict` in types for Py3.8 support

pythonhubdev commented on examples/with-litestar/with-thirdpartyemailpassword/README.md:
````diff
@@ -0,0 +1,33 @@
+## Start web app
+
+```bash
+uvicorn main:app --reload --host=0.0.0.0 --port=3001
````
Not required

pythonhubdev commented on supertokens_python/framework/litestar/framework.py:
```diff
@@ -0,0 +1,30 @@
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
```
Removed

namsnath commented on tests/litestar/test_litestar.py:
```diff
@@ -418,8 +418,8 @@ async def test_login_handle(driver_config_client: TestClient[Litestar]):
     )
-    result_dict = json.loads(response_2.content)
-    assert "s" in result_dict
+    result_Dict = json.loads(response_2.content)
```
Unintentional refactor?

namsnath commented on tests/litestar/test_litestar.py:
```diff
@@ -484,9 +484,9 @@ async def email_exists_get(
             user_context: Dict[str, Any],
         ):
-            response_dict = {"custom": True}
+            response_Dict = {"custom": True}
```
Unintentional refactor?

namsnath commented on tests/litestar/test_litestar.py:
```diff
@@ -484,9 +484,9 @@ async def email_exists_get(
+            response_Dict = {"custom": True}
             api_options.response.set_status_code(203)
-            api_options.response.set_json_content(response_dict)
+            api_options.response.set_json_content(response_Dict)
```
Unintentional refactor?

namsnath commented on tests/litestar/test_litestar.py:
```diff
@@ -514,9 +514,9 @@ async def email_exists_get(
     )
 
-    dict_response = json.loads(response.text)
+    Dict_response = json.loads(response.text)
```
Unintentional refactor?

namsnath commented on tests/litestar/test_litestar.py:
```diff
@@ -514,9 +514,9 @@ async def email_exists_get(
+    Dict_response = json.loads(response.text)
     assert response.status_code == 203
-    assert dict_response["custom"]
+    assert Dict_response["custom"]
```
Unintentional refactor?

namsnath commented on tests/litestar/test_litestar.py:
```diff
@@ -539,9 +539,9 @@ async def test_optional_session(driver_config_client: TestClient[Litestar]):
     )
 
-    dict_response = json.loads(response.text)
+    Dict_response = json.loads(response.text)
```
Unintentional refactor?

namsnath commented on tests/litestar/test_litestar.py:
```diff
@@ -539,9 +539,9 @@ async def test_optional_session(driver_config_client: TestClient[Litestar]):
+    Dict_response = json.loads(response.text)
     assert response.status_code == 200
-    assert dict_response["s"] == "empty session"
+    assert Dict_response["s"] == "empty session"
```
Unintentional refactor?

pythonhubdev replied:
Yes this was unintentional refactor will change all this.

pythonhubdev commented:
Hi, I may require some help in this facing various issues and things are failing

pythonhubdev commented:
Hi @namsnath cleanup the functions and added plugin instead of middleware to mount `supertokens` as `asgi` app to `Litestar` instead of middleware as per guidance from the `Litestar` team. Can you please review the code? And also can you help me how can I run the tests from `supertokens-auth-react` and `supertokens-website` in my local?

pythonhubdev commented:
Yeah the discussion was on discord: https://discord.com/channels/919193495116337154/1433816355177496596

pythonhubdev commented:
> Hi @namsnath cleanup the functions and added plugin instead of middleware to mount `supertokens` as `asgi` app to `Litestar` instead of middleware as per guidance from the `Litestar` team. Can you please review the code? And also can you help me how can I run the tests from `supertokens-auth-react` and `supertokens-website` in my local?

@namsnath can you help me with this please?

pythonhubdev commented:
@namsnath  I have rebased with branch 0.30 and pushed. Please check. Also I am not able to add / update the label.

namsnath commented:
Please look at the other changes required for the changed action as well. 
There will be changes required for the params and how the outputs are used. 

Approving a workflow run to see if the rest of the tests run fine.

pythonhubdev commented:
Hi I was using a relative import that was causing the issue I guess I have fixed it. Can you please check the latest commit if the changes are fine? @namsnath
